packages/js/product-editor/src/blocks/product-fields/product-details-section-description/edit.tsx

```diff
@@ -0,0 +1,360 @@
+					onClose();
+					return;
+				}
+
+				await validate( productTemplate.productData );
+
+				await editEntityRecord(
+					'postType',
+					'product',
+					productId,
+					productTemplate.productData
+				);
+
+				await saveEditedEntityRecord< Product >(
+					'postType',
+					'product',
+					productId,
+					{
+						throwOnError: true,
+					}
+				);
+
+				createSuccessNotice(
+					__( 'Product type changed.', 'woocommerce' )
+				);
+			} catch ( error ) {
+				const message = getProductErrorMessage( error as WPError );
+				createErrorNotice( message );
+			}
+
+			onClose();
+		};
+	}
+
+	function resolveIcon( iconName?: string | null ) {
+		if ( ! iconName || ! ( iconName in icons ) ) return undefined;
+
+		const { Icon, [ iconName as never ]: icon } = icons;
+
+		return <Icon icon={ icon } size={ 24 } />;
+	}
+
+	function getMenuItem( onClose: () => void ) {
+		return function renderMenuItem( productTemplate: ProductTemplate ) {
+			const isSelected =
+				selectedProductTemplate?.id === productTemplate.id;
+			return (
+				<MenuItem
+					key={ productTemplate.id }
+					info={ productTemplate.description ?? undefined }
+					isSelected={ isSelected }
+					icon={
+						isSelected
+							? resolveIcon( 'check' )
+							: resolveIcon( productTemplate.icon )
+					}
+					iconPosition="left"
+					role="menuitemradio"
+					onClick={ menuItemClickHandler( productTemplate, onClose ) }
+					className={ classNames( {
+						'components-menu-item__button--selected': isSelected,
+					} ) }
+				>
+					{ productTemplate.title }
+				</MenuItem>
+			);
+		};
+	}
+
+	async function handleModelChangeClick() {
+		try {
+			if ( isSaving ) return;
+
+			await validate( unsupportedProductTemplate?.productData );
+
+			const product = ( await saveEditedEntityRecord< Product >(
+				'postType',
+				'product',
+				productId,
+				{
+					throwOnError: true,
+				}
+			) ) ?? { id: productId };
+
+			// Avoiding to save some changes that are not supported by the current product template.
+			// So in this case those changes are saved directly to the server.
+			await saveEntityRecord(
+				'postType',
+				'product',
+				{
+					...product,
+					...unsupportedProductTemplate?.productData,
+				},
+				// @ts-expect-error Expected 3 arguments, but got 4.
+				{
+					throwOnError: true,
+				}
+			);
+
+			createSuccessNotice( __( 'Product type changed.', 'woocommerce' ) );
+
+			// Let the server manage the redirection when the product is not supported
+			// by the product editor.
+			window.location.href = getNewPath( {}, `/product/${ productId }` );
+		} catch ( error ) {
+			const message = getProductErrorMessage( error as WPError );
+			createErrorNotice( message );
+		}
+	}
+
+	return (
+		<Fill name={ rootClientId }>
+			<div { ...blockProps }>
+				<p>
+					{ createInterpolateElement(
+						/* translators: <ProductTemplate />: the product template. */
+						__( 'This is a <ProductTemplate />.', 'woocommerce' ),
```

Comment:
Hmm... we run into grammar issues with this...

For example, if you select "Affiliate product", the correct grammar would be:

"This is **an** affiliate product."

@j111q Any thoughts on how we could handle this, from a design perspective, to ensure that we always have correct grammar? One obvious solution would be:

"Product type: Affiliate product (change)"

But, some may argue that doesn't look as nice! (I like it.)